### DISCUSSION TITLE
Realtime delete event is not containing all old data when RLS is enabled

### DISCUSSION THREAD

**Discussion by AurelieV** [supabase/supabase#12471] [Questions] [answered]
Hi all,
I have a penalities table with Realtime setup. I also set up the replica full on this table, as I need the old data to update my object on my web page.
Everything was working fine, until I recently add some RLS rules on this table.
Rules are
* Only admin can delete a penality
* Only authenticated user can insert and read penality
* No one can update a penality
* 
<img width="780" alt="Capture" src="https://user-images.githubusercontent.com/5425038/219311594-27a82f88-2857-4197-91b8-f5505f5631d0.PNG">
<img width="581" alt="Capture2" src="https://user-images.githubusercontent.com/5425038/219311595-1293b3cb-fe28-4c95-8a2b-5753de419adb.PNG">

When the RLS is active, old data on the delete event only contains the primary id, and nothing else. When i disabled RLS, i am receiving the full data.

I have temporaly disabled RLS on this table so that the feature can work, but can you please explain to me how i can fix this?

**Comment by dongnez** [comment]
I have the same problem :(

**Comment by GaryAustin1** [comment, marked as answer]
This is by design.  Delete events can't use RLS so all delete operations will be sent to all users monitoring.  Because of this if you have RLS on the realtime code does not want to leak information a user may not by RLS be allowed to see.
https://supabase.com/docs/guides/realtime/postgres-changes?queryGroups=database-method&database-method=sql&queryGroups=language&language=js#receiving-old-records
![image](https://github.com/user-attachments/assets/fb55b397-15d3-4a26-8830-a9477fa1c58f)

**Reply by dongnez** [reply]
Thas sad, because I need to get the delete thing to get another table Id.
So I need to disable the whole RLS to get the full item.

There is no way to get that without disabling RLS then?

**Reply by GaryAustin1** [reply]
You might be able to do a delete trigger and insert to another table.  Then monitor that table.  You would have to have a way to clean that table up and of course another connection to monitor that and do whatever you wanted to do with deletes.

**Reply by dongnez** [reply]
Thanks for the Answer @GaryAustin1, I think I will be fine without that, its not a big deal but now I know why.
Thank you

**Reply by zhtgeneral** [reply]
> Delete events can't use RLS so all delete operations will be sent to all users monitoring. 
> if you have RLS on the realtime code does not want to leak information a user may not by RLS be allowed to see.

Maybe this feature could be added?

Specifically:
table A has a list of foreign keys. The delete event wants to read the full data of A and send it off to the app, but we don't know if we have privilege to read each of the foreign keys connected to A. We recursively check the RLS on each of the foreign tables to determine if we can in fact read the full data of A.

Example:
```
A ← check if read access is given to all
├ B ← check if read access is given to all
├ C ← check if read access is given to all
│ └ C1 ← check if read access is given to all
└ Z ← check if read access is given to all
```

after checking the RLS for each foreign key and its dependencies, we are able to get the full data returned with RLS in place.
